File: Programmers/[72414] 광고 삽입/ONE/72414.java
class Solution {

    class Log {
        int start;
        int end;

        public Log(String log) {
            String[] times = log.split("-");
            this.start = stringToTime(times[0]);
            this.end = stringToTime(times[1]);
        }

        public void accumulate(int[] window) {
            for(int i = start; i < end; i++)
                window[i]++;
        }
    }

    public String solution(String play_time, String adv_time, String[] logs) {
        int play = stringToTime(play_time);
        int adv = stringToTime(adv_time);
        int[] window = new int[play];

        if(play == adv)
            return timeToString(0);

        for (String log : logs)
            new Log(log).accumulate(window);

        return timeToString(findMax(play, adv, window));
    }

    private int stringToTime(String str) {
        String[] times = str.split(":");
        int hour = Integer.parseInt(times[0]) * 3600;
        int minute = Integer.parseInt(times[1]) * 60;
        int second  = Integer.parseInt(times[2]);

        return hour + minute + second;
    }

    private String timeToString(int time) {
        return String.format("%02d", time / 3600) + ":" +
                String.format("%02d", time % 3600 / 60) + ":" +
                String.format("%02d", time % 3600 % 60);
    }

    private int findMax(int play, int adv, int[] window) {
        int index = 0;
        long max = 0, current = 0;

        for(int i = 0; i < adv; i++)
            max += window[i];

        current = max;

        for (int i = 0, j = adv; j < play; i++, j++) {
            current = current - window[i] + window[j];
            if (current > max) {
                index = i + 1;
                max = current;
            }
        }
        return index;
    }
}

File: Programmers/[72414] 광고 삽입/Johoseong/72414.py
def convert_time(time):
    h, m, s = time.split(':')
    return int(h) * 3600 + int(m) * 60 + int(s)

def convert_str(time):
    h = time // 3600
    h = '0' + str(h) if h < 10 else str(h)
    time = time % 3600
    m = time // 60
    m = '0' + str(m) if m < 10 else str(m)
    time = time % 60
    s = '0' + str(time) if time < 10 else str(time)
    return h + ':' + m + ':' + s

def solution(play_time, adv_time, logs):
    play_time = convert_time(play_time)
    adv_time = convert_time(adv_time)               
    all_time = [0 for _ in range(play_time + 1)]

    for l in logs:
        start, end = l.split('-')
        start = convert_time(start)
        end = convert_time(end)
        all_time[start] += 1
        all_time[end] -= 1

    for i in range(1, len(all_time)): # 구간 별
        all_time[i] = all_time[i] + all_time[i - 1]

    for i in range(1, len(all_time)): # 누적
        all_time[i] = all_time[i] + all_time[i - 1]

    most_view = 0
    max_time = 0                          
    for i in range(adv_time - 1, play_time): # 누적 대비 시청자수 max 찾기
        if i >= adv_time:
            if most_view < all_time[i] - all_time[i - adv_time]:
                most_view = all_time[i] - all_time[i - adv_time]
                max_time = i - adv_time + 1
        else:
            if most_view < all_time[i]:
                most_view = all_time[i]
                max_time = i - adv_time + 1

    return convert_str(max_time)

print(solution("02:03:55", "00:14:15", ["01:20:15-01:45:14", "00:25:50-00:48:29", "00:40:31-01:00:00", "01:37:44-02:02:30", "01:30:59-01:53:29"]))